migration/bucket_migrator_test.go

```diff
@@ -0,0 +1,84 @@
+package migration_test
+
+import (
+	"bytes"
+	"errors"
+	"testing"
+
+	"github.com/NethermindEth/juno/db"
+	"github.com/NethermindEth/juno/db/pebble"
+	"github.com/NethermindEth/juno/migration"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBucketMover(t *testing.T) {
+	beforeCalled := false
+	sourceBucket := db.Bucket(0)
+	destBucket := db.Bucket(1)
+	mover := migration.NewBucketMover(sourceBucket, destBucket).WithBefore(func() {
+		beforeCalled = true
+	}).WithBatchSize(2).WithKeyFilter(func(b []byte) (bool, error) {
+		return len(b) > 1, nil
+	})
+
+	testDB := pebble.NewMemTest()
+	require.NoError(t, testDB.Update(func(txn db.Transaction) error {
+		for i := byte(0); i < 10; i++ {
+			if err := txn.Set(sourceBucket.Key([]byte{i}), []byte{i}); err != nil {
+				return err
+			}
+		}
+		return txn.Set(sourceBucket.Key(), []byte{44})
+	}))
+
+	mover.Before()
+	require.True(t, beforeCalled)
+
+	txnBudget := 100
+	for {
+		var migrationErr error
+		dbErr := testDB.Update(func(txn db.Transaction) error {
+			migrationErr = mover.Migrate(txn)
+			if errors.Is(migrationErr, migration.ErrCallWithNewTransaction) {
+				return nil // Run the migration again with a new transaction.
+			}
+			return migrationErr
+		})
+		require.NoError(t, dbErr)
+		if migrationErr != nil {
+			require.ErrorIs(t, migrationErr, migration.ErrCallWithNewTransaction)
+		} else {
+			break
+		}
+		txnBudget--
+		require.True(t, txnBudget > 0)
+	}
+
+	err := testDB.View(func(txn db.Transaction) error {
+		err := txn.Get(sourceBucket.Key(), func(b []byte) error {
+			if !bytes.Equal(b, []byte{44}) {
+				return errors.New("shouldnt have changed")
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		for i := byte(0); i < 10; i++ {
+			err = txn.Get(destBucket.Key([]byte{i}), func(b []byte) error {
+				if !bytes.Equal(b, []byte{i}) {
+					return errors.New("shouldve moved")
+				}
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+
+			require.ErrorIs(t, db.ErrKeyNotFound, txn.Get(sourceBucket.Key([]byte{i}), func(b []byte) error { return nil }))
```

Comment:
nit: can we expand this for readability?

Reply:
Done